src/pallene/ir.lua

```diff
@@ -258,136 +265,117 @@ function ir.get_dsts(cmd)
+    local cmd = block.cmds[#block.cmds]
+    if not cmd or cmd._tag ~= "ir.Cmd.Jmp" then
+        return false
+    else
+        return cmd
+    end
+end
 
-        local tag = cmd._tag
-        if     tag == "ir.Cmd.Seq" then
-            for _, c in ipairs(cmd.cmds) do
-                go(c)
-            end
-        elseif tag == "ir.Cmd.If" then
-            go(cmd.then_)
-            go(cmd.else_)
-        elseif tag == "ir.Cmd.Loop" then
-            go(cmd.body)
-        elseif tag == "ir.Cmd.For" then
-            go(cmd.body)
+function ir.get_jmpIfFalse(block)
+    if #block.cmds < 2 then
+        return false
+    else
+        local cmd = block.cmds[#block.cmds - 1]
+        if cmd._tag == "ir.Cmd.JmpIfFalse" then
+            return cmd
         else
-            -- no recursion needed
+            return false
         end
     end
-
-    return coroutine.wrap(function()
-        go(root_cmd)
-    end)
 end
 
-function ir.flatten_cmd(root_cmd)
-    local res = {}
-    for cmd in ir.iter(root_cmd) do
-        table.insert(res, cmd)
-    end
-    return res
+local function remove_jmpIfFalse(block)
+    local jmp_false_index = #block.cmds - 1
+    local jmp_false = block.cmds[jmp_false_index]
+    assert(jmp_false)
+    local jmp = block.cmds[#block.cmds]
+    assert(jmp)
+    block.cmds[jmp_false_index] = jmp
+    table.remove(block.cmds)
 end
 
--- Transform an ir.Cmd, via a mapping function that modifies individual nodes.
--- Returns the new root node. Child nodes are modified in-place.
--- If the mapping function returns a falsy value, the original version of the node is kept.
-function ir.map_cmd(root_cmd, f)
-    local function go(cmd)
-        -- Transform child nodes recursively
-        local tag = cmd._tag
-        if     tag == "ir.Cmd.Seq" then
-            for i = 1, #cmd.cmds do
-                cmd.cmds[i] = go(cmd.cmds[i])
+-- Returns list of block indices. Unreacheable blocks won't appear on the list, which means the
+-- returned list might be smaller than the block list.
+function ir.get_depth_search_topological_sort(block_list)
+    local order = {}
+    local visited = {}
+    for i,_ in ipairs(block_list) do
+        visited[i] = false
+    end
+    local function depth_search(block_i)
+        if not visited[block_i] then
+            visited[block_i] = true
+            local block = block_list[block_i]
+            local jmp = ir.get_jmp(block)
+            local jmp_false = ir.get_jmpIfFalse(block)
+            if jmp  then
+                depth_search(jmp.target)
             end
-        elseif tag == "ir.Cmd.If" then
-            cmd.then_ = go(cmd.then_)
-            cmd.else_ = go(cmd.else_)
-        elseif tag == "ir.Cmd.Loop" then
-            cmd.body = go(cmd.body)
-        elseif tag == "ir.Cmd.For" then
-            cmd.body = go(cmd.body)
-        else
-            -- no child nodes
+            if jmp_false then
+                depth_search(jmp_false.target)
+            end
```

Comment:
These get_jump funcitons feel strange. Can we avoid them by giving a uniform interface to the jump instructions? E.g. `block[#block].target`

Reply:
This would be a problem for the last block, it has no commands. The cleaner solution I could find for avoiding  the get_jump was to not have jump instructions and let the jumps be fields on the BasicBlock struct. In that case we would have to do this thing of putting the boolean source of the conditional jump as a special command at the end of the block, but then it would be a more uniform format because this special command will appear if and only if the block has a conditional jump.